### PR TITLE
Reduce memory usage when scheduler start jobs

### PR DESCRIPTION
**Pull request by nickva**
Scheduler calls pending_jobs which returns all available jobs. This could be
a huge list of all job records. Instead use a smaller list of tuples, sort
and filter those then start jobs based on job ids.

### CODE REVIEW

File: src/couch_replicator_scheduler.erl [hunk 1]
```diff
@@ -381,9 +387,19 @@ pending_job_count() ->
...
 
 
--spec pending_jobs() -> [#job{}].
-pending_jobs() ->
-    ets:match_object(?MODULE, #job{pid=undefined, _='_'}).
+% Get list of not recently crashed job infos {#job.id, last_started(Job)}.
+% A smaller tuple is used here to reduce memory usage due to returning
+% potentially a huge list of pending job records.
+-spec pending_jobs_info() -> [{job_id(), erlang:timestamp()}].
+pending_jobs_info() ->
```
Comment (sagelywizard) [inline src/couch_replicator_scheduler.erl]: What do you think about passing in `Count` to `pending_jobs_info` and only accumulating the top `Count` entries? This still accumulates an element for ~every entry in the table.

Comment (nickva) [inline src/couch_replicator_scheduler.erl]: @sagelywizard Great idea. let's try that. Perhaps have a sorted set (`gb_sets`). If its size has grown above the `Count` and current element is older than youngest one. Add element in set and remove youngest. Take will take more CPU time but code might look cleaner and memory should look better as well.

**Comment by nickva** [comment]
@rnewson @sagelywizard Hot-patched this code a test cluster. So far see memory size reduction from 800MB to 1MB.

**Comment by sagelywizard** [comment]
You know, if this change as-is reduces the memory footprint by 100x, it might not be worth adding the further gb_trees-type optimization which you mentioned. Just a thought. Can talk about it on Slack tomorrow.

File: src/couch_replicator_scheduler.erl [hunk 2]
```diff
@@ -271,12 +271,54 @@ handle_config_terminate(Self, _, _) ->
...
+pending_jobs(0) ->
+    % Handle this case as user could set max_churn to 0. If this is passed to
+    % other function clause it will crash as gb_sets:largest assumes set is not
+    % empty.
+    [];
+
+pending_jobs(Count) when is_integer(Count), Count > 0 ->
+    Set0 = gb_sets:new(),  % [{LastStart, Job},...]
+    Now = os:timestamp(),
+    {Set1, _, _} = ets:foldl(fun pending_fold/2, {Set0, Now, Count}, ?MODULE),
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl:286]: a record instead of the 3-tuple would be more readable.

Comment (rnewson) [inline src/couch_replicator_scheduler.erl:286]: discuss.

File: src/couch_replicator_scheduler.erl [hunk 3]
```diff
@@ -271,12 +271,54 @@ handle_config_terminate(Self, _, _) ->
...
+
+
+pending_fold(Job, {Set, Now, Count}) ->
+    Set1 = case {not_recently_crashed(Job, Now), gb_sets:size(Set) >= Count} of
+        {true, true} ->
+             % Job is healthy but already reached accumulated limit, so might
+             % have to replace one of the accumulated jobs
+             pending_maybe_replace(Job, Set);
+        {true, false} ->
+             % Jobs is healthy and but haven't reached the limit, so add job
```
Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: "Job"

Comment (rnewson) [inline src/couch_replicator_scheduler.erl]: s/and but/and we/

**Comment by rnewson** [comment]
+1

**Pull request merged**